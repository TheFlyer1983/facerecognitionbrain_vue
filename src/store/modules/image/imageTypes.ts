export interface ImageState {
  input: string;
  imageUrl: string;
  boxes: Array<Box>;
}

export interface Box {
  id: string;
  region_info: Region;
  data: Array<Concept>;
  value: number;
}

export interface Region {
  bounding_box: BoundingBox;
}

export interface BoundingBox {
  top_row: number;
  left_col: number;
  bottom_row: number;
  right_col: number;
}

export interface Concept {
  id: string;
  name: string;
  value: number;
  app_id: string;
}
