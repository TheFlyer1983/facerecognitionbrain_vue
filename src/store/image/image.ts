import { defineStore } from 'pinia';
import { useUserStore } from '../user';
import request from '@/functions/request';
import { ImageState } from './imageTypes';
import { endpoints } from '@/constants';

export const useImageStore = defineStore('image', {
  state: (): ImageState => ({
    imageUrl: '',
    boxes: []
  }),

  actions: {
    async submitURL() {
      try {
        const response = await request.post(endpoints.imageURL, {
          input: this.imageUrl
        });

        this.boxes = response.data;
        this.increaseEntries();
      } catch (error) {
        console.error(error);
      }
    },

    async increaseEntries() {
      const userStore = useUserStore();

      if (!userStore.user) return;

      try {
        const response = await request.put(endpoints.image, {
          id: userStore.user.id
        });

        userStore.user.entries = response.data.entries;
        userStore.getRank();
      } catch (error) {
        console.error(error);
      }
    }
  }
});
