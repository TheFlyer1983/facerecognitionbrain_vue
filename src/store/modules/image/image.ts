import { endpoints } from '@/constants';
import { UserState } from '../user/userTypes';
import { ImageActionContext, ImageState } from './imageTypes';
import request from '@/functions/request';

export const state: ImageState = {
  imageUrl: '',
  boxes: []
};

export const getters = {
  getImageURL(state: ImageState): ImageState['imageUrl'] {
    return state.imageUrl;
  },

  getBoxes(state: ImageState): ImageState['boxes'] {
    return state.boxes;
  }
};

export const mutations = {
  setImageURL(state: ImageState, payload: ImageState['imageUrl']): void {
    state.imageUrl = payload;
  },

  setBoxes(state: ImageState, payload: ImageState['boxes']): void {
    state.boxes = payload;
  }
};

export const actions = {
  async submitURL({
    state,
    commit,
    dispatch
  }: ImageActionContext): Promise<void> {
    try {
      const response = await request.post(endpoints.imageURL, {
          input: state.imageUrl
        });

      commit('setBoxes', response.data);
      dispatch('increaseEntries');
    } catch (error) {
      console.error(error);
    }
  },

  async increaseEntries({
    commit,
    dispatch,
    rootGetters
  }: ImageActionContext): Promise<void> {
    const user = rootGetters['user/getUser'] as UserState['user'];

    try {
      const response = await request.put(endpoints.image, {
          id: user?.id
        });

      commit('user/updateEntries', response.data.entries, { root: true });
      dispatch('user/getRank', null, { root: true });
    } catch (error) {
      console.error(error);
    }
  }
};

const imageModule = {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
};

export default imageModule;
