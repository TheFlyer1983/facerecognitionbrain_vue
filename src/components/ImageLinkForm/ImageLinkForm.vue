<script setup lang="ts">
import { getters } from '@/store/modules/image/image';
import { computed } from 'vue';
import { useStore } from 'vuex';

const store = useStore();

const imageURL = computed<ReturnType<typeof getters['getImageURL']>>({
  get: () => store.getters['image/getImageURL'],
  set: (value) => store.commit('image/setImageURL', value)
});

async function submitURL() {
  store.dispatch('image/submitURL')
}
</script>

<template>
  <div class="image-link">
    <p class="image-link--text">
      This Magic Brain will detect faces in your pictures. Give it a try.
    </p>
    <div class="image-link-wrapper">
      <div class="image-link-wrapper-form">
        <input
          type="text"
          v-model="imageURL"
          class="image-link-wrapper-form--input"
        />
        <button @click="submitURL" class="image-link-wrapper-form--button" data-test="submit">Detect</button>
      </div>
    </div>
  </div>
</template>

<style lang="scss" scoped>
.image-link {
  &--text {
    font-size: 1.5rem;
  }

  &-wrapper {
    margin-right: auto;
    margin-left: auto;

    &-form {
      display: flex;
      margin-right: auto;
      margin-left: auto;
      width: 700px;
      background: radial-gradient(
            circle farthest-side at 0% 50%,
            #fb1 23.5%,
            rgba(240, 166, 17, 0) 0
          )
          21px 30px,
        radial-gradient(
            circle farthest-side at 0% 50%,
            #b71 24%,
            rgba(240, 166, 17, 0) 0
          )
          19px 30px,
        linear-gradient(
            #fb1 14%,
            rgba(240, 166, 17, 0) 0,
            rgba(240, 166, 17, 0) 85%,
            #fb1 0
          )
          0 0,
        linear-gradient(
            150deg,
            #fb1 24%,
            #b71 0,
            #b71 26%,
            rgba(240, 166, 17, 0) 0,
            rgba(240, 166, 17, 0) 74%,
            #b71 0,
            #b71 76%,
            #fb1 0
          )
          0 0,
        linear-gradient(
            30deg,
            #fb1 24%,
            #b71 0,
            #b71 26%,
            rgba(240, 166, 17, 0) 0,
            rgba(240, 166, 17, 0) 74%,
            #b71 0,
            #b71 76%,
            #fb1 0
          )
          0 0,
        linear-gradient(90deg, #b71 2%, #fb1 0, #fb1 98%, #b71 0%) 0 0 #fb1;
      background-size: 40px 60px;
      padding: 2rem;
      border-radius: 0.5rem;
      box-shadow: 4px 4px 8px 0px rgba(0, 0, 0, 0.2);

      &--input {
        margin-left: auto;
        margin-right: auto;
        padding: 0.5rem;
        font-size: 1.25rem;
        width: 70%;
      }

      &--button {
        cursor: pointer;
        width: 30%;
        font-size: 1.25rem;
        background-color: var(--light-purple);
        padding: var(--spacing-small);
        color: white;
        text-decoration: none;
        transition: color 0.15s ease-in;
        -moz-osx-font-smoothing: grayscale;
        backface-visibility: hidden;
        transform: translateZ(0);
        transition: transform 0.25s ease-out;

        &:hover {
          transform: scale(1.05);
        }
      }
    }
  }
}
</style>
