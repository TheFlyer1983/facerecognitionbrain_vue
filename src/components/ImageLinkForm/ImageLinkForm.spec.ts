import { shallowMount, VueWrapper } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import { useImageStore } from '@/store/modules/image';
import ImageLinkForm from './ImageLinkForm.vue';

const mockedImageURL =
  'https://images.unsplash.com/photo-1597223557154-721c1cecc4b0?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MjB8fGZhY2V8ZW58MHx8MHx8&w=1000&q=80';

const pinia = createTestingPinia();
const mockImageStore = useImageStore(pinia);

describe('Given the `ImageLinkForm` component', () => {
  const render = () =>
    shallowMount(ImageLinkForm, { global: { plugins: [pinia] } });
  let wrapper: VueWrapper<InstanceType<typeof ImageLinkForm>>;

  describe('when it is rendered', () => {
    beforeEach(() => {
      wrapper = render();
    });

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('when a new image is input', () => {
      beforeEach(() => {
        wrapper.find('input').setValue(mockedImageURL);
      });

      it('should update the state correctly', () => {
        expect(mockImageStore.imageUrl).toStrictEqual(mockedImageURL);
      });
    });

    describe('when the submit button is clicked', () => {
      beforeEach(() => {
        wrapper.find('[data-test="submit"]').trigger('click');
      });

      it('should dispatch the correct action', () => {
        expect(mockImageStore.submitURL).toHaveBeenCalled();
      });
    });
  });
});
