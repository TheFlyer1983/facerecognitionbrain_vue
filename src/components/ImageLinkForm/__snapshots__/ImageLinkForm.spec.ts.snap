// Vitest Snapshot v1

exports[`Given the \`ImageLinkForm\` component > when it is rendered > should match the snapshot 1`] = `
<div
  class="image-link"
  data-v-1cdfb4fa=""
>
  <p
    class="image-link--text"
    data-v-1cdfb4fa=""
  >
     This Magic Brain will detect faces in your pictures. Give it a try. 
  </p>
  <div
    class="image-link-wrapper"
    data-v-1cdfb4fa=""
  >
    <div
      class="image-link-wrapper-form"
      data-v-1cdfb4fa=""
    >
      <input
        class="image-link-wrapper-form--input"
        data-v-1cdfb4fa=""
        type="text"
      />
      <button
        class="image-link-wrapper-form--button"
        data-test="submit"
        data-v-1cdfb4fa=""
      >
        Detect
      </button>
    </div>
  </div>
</div>
`;
